pkg/xray/watches.go

```diff
@@ -0,0 +1,246 @@
+package xray
+
+import (
+	"context"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"log"
+	"net/http"
+
+	"github.com/go-resty/resty/v2"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type WatchGeneralData struct {
+	Name        string `json:"name"`
+	Description string `json:"description,omitempty"`
+	Active      bool   `json:"active,omitempty"`
+}
+
+type WatchFilterValue struct {
+	Key   string `json:"key,omitempty"`
+	Value string `json:"value,omitempty"`
+}
+
+type WatchFilter struct {
+	Type  string `json:"type,omitempty"`
+	Value string `json:"value,omitempty"`
+}
+
+type WatchProjectResource struct {
+	Type            string        `json:"type,omitempty"`
+	BinaryManagerId string        `json:"bin_mgr_id,omitempty"`
+	Filters         []WatchFilter `json:"filters,omitempty"`
+	// Watch a repo
+	Name string `json:"name,omitempty"`
+}
+
+type WatchProjectResources struct {
+	Resources []WatchProjectResource `json:"resources,omitempty"`
+}
+
+type WatchAssignedPolicy struct {
+	Name string `json:"name,omitempty"`
+	Type string `json:"type,omitempty"`
+}
+
+type Watch struct {
+	GeneralData      WatchGeneralData      `json:"general_data,omitempty"`
+	ProjectResources WatchProjectResources `json:"project_resources,omitempty"`
+	AssignedPolicies []WatchAssignedPolicy `json:"assigned_policies,omitempty"`
+}
+
+func unpackWatch(d *schema.ResourceData) *Watch {
+	watch := new(Watch)
+
+	gd := WatchGeneralData{
+		Name: d.Get("name").(string),
+	}
+	if v, ok := d.GetOk("description"); ok {
+		gd.Description = v.(string)
+	}
+	if v, ok := d.GetOk("active"); ok {
+		gd.Active = v.(bool)
+	}
+	watch.GeneralData = gd
+
+	pr := WatchProjectResources{}
+	if v, ok := d.GetOk("watch_resource"); ok {
+		r := []WatchProjectResource{}
+		for _, res := range v.([]interface{}) {
+			r = append(r, unpackProjectResource(res))
+		}
+		pr.Resources = r
+	}
+	watch.ProjectResources = pr
+
+	ap := []WatchAssignedPolicy{}
+	if v, ok := d.GetOk("assigned_policy"); ok {
+		for _, pol := range v.([]interface{}) {
+			ap = append(ap, unpackAssignedPolicy(pol))
+		}
+	}
+	watch.AssignedPolicies = ap
+
+	return watch
+}
+
+func unpackProjectResource(rawCfg interface{}) WatchProjectResource {
+	resource := new(WatchProjectResource)
+
+	cfg := rawCfg.(map[string]interface{})
+	resource.Type = cfg["type"].(string)
+
+	if v, ok := cfg["bin_mgr_id"]; ok {
+		resource.BinaryManagerId = v.(string)
+	}
+	if v, ok := cfg["name"]; ok {
+		resource.Name = v.(string)
+	}
+
+	if v, ok := cfg["filter"]; ok {
+		resourceFilters := unpackFilters(v.([]interface{}))
+		resource.Filters = resourceFilters
+	}
+
+	return *resource
+}
+
+func unpackFilters(list []interface{}) []WatchFilter {
+	var filters []WatchFilter
+
+	for _, raw := range list {
+		filter := new(WatchFilter)
+		f := raw.(map[string]interface{})
+		filter.Type = f["type"].(string) // TODO: recognize the type of the filter
```

Comment:
What does this comment mean?

Reply:
The filter will have one more type, where the value will be not a string, but a JSON object. It's not implemented yet